EventFilter/EcalRawToDigi/plugins/UnpackGPU.cu

```diff
@@ -0,0 +1,340 @@
+          ElectronicsIdGPU eid{fed2dcc(fed), ttid, stripid, xtalid};
+          auto const didraw = isBarrel ? compute_ebdetid(eid) : eid2did[eid.linearIndex()];
+          // FIXME: what kind of channels are these guys
+          if (didraw == 0)
+            continue;
+
+          // get samples
+          uint16_t sampleValues[10];
+          sampleValues[0] = (wdata >> 16) & 0x3fff;
+          sampleValues[1] = (wdata >> 32) & 0x3fff;
+          sampleValues[2] = (wdata >> 48) & 0x3fff;
+          auto const wdata1 = current_tower_block[2 + i_to_access * 3];
+          sampleValues[3] = wdata1 & 0x3fff;
+          sampleValues[4] = (wdata1 >> 16) & 0x3fff;
+          sampleValues[5] = (wdata1 >> 32) & 0x3fff;
+          sampleValues[6] = (wdata1 >> 48) & 0x3fff;
+          auto const wdata2 = current_tower_block[3 + i_to_access * 3];
+          sampleValues[7] = wdata2 & 0x3fff;
+          sampleValues[8] = (wdata2 >> 16) & 0x3fff;
+          sampleValues[9] = (wdata2 >> 32) & 0x3fff;
+          //printf("stripid = %u xtalid = %u\n", stripid, xtalid);
+
+          // check gain
+          bool isSaturation = true;
+          short firstGainZeroSampID{-1}, firstGainZeroSampADC{-1};
+          for (uint32_t si = 0; si < 10; si++) {
+            if (gainId(sampleValues[si]) == 0) {
+              firstGainZeroSampID = si;
+              firstGainZeroSampADC = adc(sampleValues[si]);
+              break;
+            }
+          }
+          if (firstGainZeroSampID != -1) {
+            unsigned int plateauEnd = std::min(10u, (unsigned int)(firstGainZeroSampID + 5));
+            for (unsigned int s = firstGainZeroSampID; s < plateauEnd; s++) {
+              if (gainId(sampleValues[s]) == 0 && adc(sampleValues[s]) == firstGainZeroSampADC) {
+                ;
+              } else {
+                isSaturation = false;
+                break;
+              }  //it's not saturation
+            }
+            // get rid of channels which are stuck in gain0
+            if (firstGainZeroSampID < 3) {
+              isSaturation = false;
+            }
+            if (!isSaturation)
+              continue;
+          } else {  // there is no zero gainId sample
+            // gain switch check
+            short numGain = 1;
+            bool gainSwitchError = false;
+            for (unsigned int si = 1; si < 10; si++) {
+              if ((gainId(sampleValues[si - 1]) > gainId(sampleValues[si])) && numGain < 5)
+                gainSwitchError = true;
+              if (gainId(sampleValues[si - 1]) == gainId(sampleValues[si]))
+                numGain++;
+              else
+                numGain = 1;
+            }
+            if (gainSwitchError)
+              continue;
+          }
+
+          auto const pos = atomicAdd(pChannelsCounter, 1);
+
+          // store to global
+          ids[pos] = didraw;
+          samples[pos * 10] = sampleValues[0];
+          samples[pos * 10 + 1] = sampleValues[1];
```

Comment:
why not a for loop? do you need to unroll manually? If so, make a comment here in the code to that effect.

Reply:
I will not, but feel free to make a PR with such comment once this has been merged.